apis/clusters/v1beta1/validation_test.go

```diff
@@ -0,0 +1,1063 @@
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := validateTwoFactorDelete(tt.args.new, tt.args.old); (err != nil) != tt.wantErr {
+				t.Errorf("validateTwoFactorDelete() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_validateIngestNodes(t *testing.T) {
+	type args struct {
+		new []*OpenSearchIngestNodes
+		old []*OpenSearchIngestNodes
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "change len of the IngestNodes",
+			args: args{
+				new: []*OpenSearchIngestNodes{{}},
+				old: []*OpenSearchIngestNodes{},
+			},
+			wantErr: true,
+		},
+		{
+			name: "change field of the IngestNode",
+			args: args{
+				new: []*OpenSearchIngestNodes{{
+					NodeSize:  "new",
+					NodeCount: 1,
+				}},
+				old: []*OpenSearchIngestNodes{{
+					NodeSize:  "test",
+					NodeCount: 1,
+				}},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := validateIngestNodes(tt.args.new, tt.args.old); (err != nil) != tt.wantErr {
+				t.Errorf("validateIngestNodes() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_validateClusterManagedNodes(t *testing.T) {
+	type args struct {
+		new []*ClusterManagerNodes
+		old []*ClusterManagerNodes
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "change len of the ClusterManagerNodes",
+			args: args{
+				new: []*ClusterManagerNodes{{}},
+				old: []*ClusterManagerNodes{},
+			},
+			wantErr: true,
+		},
+		{
+			name: "change field of the ClusterManagerNode",
+			args: args{
+				new: []*ClusterManagerNodes{{
+					NodeSize:         "new",
+					DedicatedManager: false,
+				}},
+				old: []*ClusterManagerNodes{{
+					NodeSize:         "test",
+					DedicatedManager: false,
+				}},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := validateClusterManagedNodes(tt.args.new, tt.args.old); (err != nil) != tt.wantErr {
+				t.Errorf("validateClusterManagedNodes() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func Test_validateTagsUpdate(t *testing.T) {
+	type args struct {
+		new map[string]string
+		old map[string]string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "change len of the Tags",
+			args: args{
+				new: map[string]string{"": ""},
+				old: map[string]string{},
+			},
+			wantErr: true,
+		},
+		{
+			name: "change tag",
+			args: args{
+				new: map[string]string{"tag1": "new"},
+				old: map[string]string{"tag1": "test"},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := validateTagsUpdate(tt.args.new, tt.args.old); (err != nil) != tt.wantErr {
+				t.Errorf("validateTagsUpdate() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
```

Comment:
Why do you have only error cases?
Could you please also add successful cases where there are none?

Reply:
it has a lot of cases when we use valid data so there are successful cases also.Als

Reply:
What do you mean? You don't have any `wantErr: false` for `Test_validateTagsUpdate` or `Test_validateClusterManagedNodes` and there are other functions

Reply:
I have `wantErr: false` in cases that validate creation on updating of resource such as `Test_validateTwoFactorDelete` or `Test_validateAppVersion`. But do we need a success cases when data is immutable?(`old` and  `new` structs are identical)

Reply:
Yea, in tests you need to handle error and successful cases by best practices, so I think it makes sense.

Reply:
agree, refactored